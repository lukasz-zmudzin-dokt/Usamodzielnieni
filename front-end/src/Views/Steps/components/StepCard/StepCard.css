@media (max-width: 768px) {
    .mb-3 {
        width: 100%;
        height: 100%;
        display: flex;
    }
}