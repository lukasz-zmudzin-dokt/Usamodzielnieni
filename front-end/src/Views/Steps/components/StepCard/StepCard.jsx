import React, { useContext } from "react";
import { Button, Card } from "react-bootstrap";
import Player from "components/Player";
import { UserContext } from "context";
import { staffTypes } from "constants/staffTypes";

const StepCard = ({ step, setCurrent, wantsDelete, wantsEdit, path }) => {
  const user = useContext(UserContext);
  return (
    <Card className="stepCard">
      <Card.Body>
        <div>
          <Player src={step.video} />
        </div>
        <Card.Text as="h4" text-muted="true">
          Opis kroku:
        </Card.Text>
        {step.description}
        {((step.next && step.next.length > 0) || step.id === path[path.length - 1]) && 
          <Card.Text className="mt-3" as="h4" text-muted="true">
            Akcje:
          </Card.Text>
        }
        {step.next?.map((child) => (
          <Button
            variant="primary"
            block
            key={child.id}
            onClick={(e) => setCurrent(child.id)}
          >
            {child.choiceName}
          </Button>
        ))}
        {path[path.length - 1] === step.id && (
          <Button
            variant="secondary"
            block
            onClick={(e) => {
              setCurrent(path[0]);
              window.scrollTo(0, 0);
            }}
          >
            Początek
          </Button>
        )}
        {path[path.length - 1] === step.id &&
        user.data.group_type?.includes(staffTypes.BLOG_MODERATOR) ? (
          <Button variant="warning" block onClick={() => wantsEdit(true)}>
            Edytuj krok
          </Button>
        ) : null}
        {path[path.length - 1] === step.id &&
        user.data.group_type?.includes(staffTypes.BLOG_MODERATOR) ? (
          <Button variant="danger" block onClick={() => wantsDelete(true)}>
            Usuń krok
          </Button>
        ) : null}
      </Card.Body>
    </Card>
  );
};

export default StepCard;
