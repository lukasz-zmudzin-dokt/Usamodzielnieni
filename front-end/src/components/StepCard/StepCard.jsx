import React, {useEffect, useState} from "react";
import proxy from "config/api";
import {Alert, Button, Card, Container} from "react-bootstrap";
import YouTube from "react-youtube";

const getCardContent = async (token, id) => {
    const url = proxy.steps + id + "/";
    const headers = {
        Authorization: "Token " + token,
        "Content-type": "application/json"
    };
    const response = await fetch(url, {method: "GET", headers});

    if (response.status === 200) {
        return await response.json();
    } else {
        throw response.status;
    }
};

const StepCard = ({stepId, token}) => {
    const [id, setId] = useState(stepId);
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState(false);
    const [disabled, setDisabled] = useState(false);
    const [content, setContent] = useState({});

    useEffect(() => {
        const loadCard = async (token) => {
            setError(false);
            setLoading(true);
            setDisabled(true);
            try {
                const res = await getCardContent(token, id);
                setContent(mapContent(res));
            } catch(e) {
                setError(true);
            } finally {
                setLoading(false);
                setDisabled(false);
            }
        };
        loadCard(token)
    }, [token, id]);

    const mapContent = (data) => ({
        id: data.id,
        title: data.title,
        video: data.video,
        description: data.description,
        type: data.type,
        nextSteps: mapSteps(data.next_steps)
    });

    const mapSteps = (steps) => {
        return steps.map(child => ({
            id: child.id,
            title: child.choice_name
        }))
    };

    const videoProps = {
        height: '390',
        width: '640',
        playerVars: {
            autoplay: 1
        }
    };

    const msg = error ? <Alert variant="danger">Wystąpił błąd podczas ładowania kroku</Alert> :
                loading ? <Alert variant="info">Ładowanie kroku</Alert> :
                undefined;


    return (
        <Container>
            <Card>
                <Card.Header>{content.title}</Card.Header>
                {msg ? <Card.Body>{msg}</Card.Body> : (
                    <Card.Body>
                        <YouTube className="mb-3" videoId={content.video || "dQw4w9WgXcQ"} opts={videoProps} onReady={this._onReady} />
                        <div>
                            <Card.Text>Opis kroku:</Card.Text>
                            {content.description}
                        </div>
                        <Card.Text>Akcje:</Card.Text>
                        {content.nextSteps.map(step => (
                            <Button variant="primary" block key={step.id} onClick={e => setId(step.id)}>{disabled ? "Ładowanie..." : step.title}</Button>
                        ))}
                    </Card.Body>
                )}
            </Card>
        </Container>
    )
};

export default StepCard;