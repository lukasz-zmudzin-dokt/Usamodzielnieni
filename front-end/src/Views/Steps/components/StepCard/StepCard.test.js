import StepCard from "./StepCard";
import React from "react";
import {render, fireEvent} from "@testing-library/react";

describe("StepCardTest", () => {
    let setCurrent, step;
    beforeEach(() => {
        step = {
            id: "2",
            type: "main",
            title:
                "Tytuł głównego kroku 2 123 123",
            description: "Opis kroku 2 wraz z filmikami.",
            video: "rickrolltylkozahaszowany",
            next: [{ id: "3", choiceName: "Dalej" }],
        };
        setCurrent = jest.fn();
    });

    it('should match snapshot', () => {
        const {container} = render(
            <StepCard step={step} setCurrent={setCurrent} />
        );

        expect(container).toMatchSnapshot();
    });

    it('should call setCurrent with right params', () => {
        const {getByText} = render(
            <StepCard step={step} setCurrent={setCurrent} />
        );

        fireEvent.click(getByText("Dalej"));
        const id = step.next.find(child => child.choiceName === "Dalej").id;

        expect(setCurrent).toHaveBeenCalledWith(id);
    });

});