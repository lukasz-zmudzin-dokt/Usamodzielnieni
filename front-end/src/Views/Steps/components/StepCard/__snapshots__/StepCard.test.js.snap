// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepCardTest should match snapshot 1`] = `
<div>
  <div
    class="stepCard card"
  >
    <div
      class="card-body"
    >
      <div>
        <div
          class=""
        >
          <div
            class="yt_player"
          />
        </div>
      </div>
      <h4
        class="card-text"
        text-muted="true"
      >
        Opis kroku:
      </h4>
      Opis kroku 2 wraz z filmikami.
      <h4
        class="mt-3 card-text"
        text-muted="true"
      >
        Akcje:
      </h4>
      <button
        class="btn btn-primary btn-block"
        type="button"
      >
        Dalej
      </button>
    </div>
  </div>
</div>
`;
